cpp/src/io/orc/stripe_data.cu

```diff
@@ -132,6 +132,146 @@ struct orcdec_state_s {
+  {
+    if (_status == status::CAN_WRITE_TO_CACHE) {
+      _run_length      = run_length;
+      _reusable_length = (_run_length > max_length) ? (_run_length - max_length) : 0;
+    }
+  }
+
+  /**
+   * @brief Adjust the maximum length allowed to be consumed when the length of the first run is
+   * greater than it.
+   *
+   * @param[in] max_length The maximum length allowed to be consumed for the DATA stream.
+   * @return A new maximum length.
+   */
+  __device__ uint32_t adjust_max_length(uint32_t max_length)
+  {
+    auto new_max_length{max_length};
+    if (_status == status::CAN_READ_FROM_CACHE and _reusable_length > 0) {
+      new_max_length -= _reusable_length;
+    }
+    return new_max_length;
+  }
+
+  /**
+   * @brief Copy the excess data from the intermediate buffer for the DATA stream to the cache.
+   *
+   * @param[in] src Intermediate buffer for the DATA stream.
+   * @param[out] cache Local variable serving as the cache for the DATA stream.
+   */
+  __device__ void write_to_cache(int64_t* src, int64_t& cache)
+  {
+    if (_status != status::CAN_WRITE_TO_CACHE) { return; }
+
+    const auto tid = threadIdx.x;
+
+    // All threads in the block always take a uniform code path for the following branches.
+    // _reusable_length ranges between [0, 512].
+    if (_reusable_length > 0) {
+      const auto length_to_skip = _run_length - _reusable_length;
+      if (tid < _reusable_length) {
+        const auto src_idx = tid + length_to_skip;
+        cache              = src[src_idx];
+      }
+      // Block until all writes are done to safely change _status.
+      __syncthreads();
+      if (tid == 0) { _status = status::CAN_READ_FROM_CACHE; }
+    } else {
+      __syncthreads();
+      if (tid == 0) { _status = status::DISABLED; }
+    }
+
+    __syncthreads();
+  }
+
+  /**
+   * @brief Copy the cached data to the intermediate buffer for the DATA stream.
+   *
+   * @param[in,out] dst Intermediate buffer for the DATA stream.
+   * @param[in,out] rle Run length decoder state object.
+   * @param[in] cache Local variable serving as the cache for the DATA stream.
+   */
+  __device__ void read_from_cache(int64_t* dst, orc_rlev2_state_s* rle, int64_t cache)
+  {
+    if (_status != status::CAN_READ_FROM_CACHE) { return; }
+
+    const auto tid = threadIdx.x;
+
+    // First, shift the data up
+    const auto dst_idx = tid + _reusable_length;
+    const auto v       = (dst_idx < rle->num_vals + _reusable_length) ? dst[tid] : 0;
+    __syncthreads();
+
+    if (dst_idx < rle->num_vals + _reusable_length) { dst[dst_idx] = v; }
+    __syncthreads();
+
+    // Second, insert the cached data
+    if (tid < _reusable_length) { dst[tid] = cache; }
+    __syncthreads();
+
+    if (tid == 0) {
+      _status = status::DISABLED;
+      rle->num_vals += _reusable_length;
+    }
+
+    __syncthreads();
+  }
+
+ private:
+  status _status;  ///< The status of the run cache manager.
+  uint32_t
```

Comment:
I chose `uint32_t` to be consistent with existing code out there. All variables that `_reusable_length` interacted with are 32-bit unsigned ints, such as `maxvals`, `threadIdx.x`, `rle->num_vals`. For size-related type, I see unsigned ints being frequently used, such as `malloc`/`memcpy` within the kernel. I know the argument [against using unsigned in favor of signed](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/#loop-counters-signed-vs-unsigned) applies to the loop counter, but always have some doubt on the actual performance impact. 

`uint16_t` would be a great option to minimize the size, but since that's not an immediate concern, I'd like to avoid some `static_cast` when interacting with 32-bit unsigned ints in existing code.